src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1271,38 +1284,90 @@ JoinRangeTableEntry(JoinExpr *joinExpr, List *dependentJobList, List *rangeTable
 	rangeTableEntry->eref->colnames = joinedColumnNames;
 	rangeTableEntry->joinaliasvars = joinedColumnVars;
 
-	SetJoinRelatedColumnsCompat(rangeTableEntry,
-								leftColumnNames, rightColumnNames, leftColumnVars,
+	SetJoinRelatedColumnsCompat(rangeTableEntry, leftRelId, rightRelId, leftColumnVars,
 								rightColumnVars);
 
 	return rangeTableEntry;
 }
 
 
+/*
+ * SetJoinRelatedColumnsCompat sets join related fields on the given range table entry.
+ */
 static void
-SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry,
-							List *leftColumnNames, List *rightColumnNames,
+SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry, Oid leftRelId, Oid rightRelId,
 							List *leftColumnVars, List *rightColumnVars)
 {
 	#if PG_VERSION_NUM >= PG_VERSION_13
 
 	/* We don't have any merged columns so set it to 0 */
 	rangeTableEntry->joinmergedcols = 0;
-	int numvars = list_length(leftColumnVars);
-	for (int varId = 1; varId <= numvars; varId++)
+
+	if (OidIsValid(leftRelId))
```

Comment:
Do we want to distinguish between RTE_RELATION vs RTE_JOIN here? If so,  can't we explicitly check that? Otherwise, reading this block is confusing

Reply:
I'm further confused.
My  understanding from PG  code is that `joinleftcols `/ `joinrightcols ` are only relevant for `RTE_JOIN`. However, when I remove `if (OidIsValid(leftRelId))` case, we hit the crash again. What am I missing here, do you have any ideas?

The fields are under [`Fields valid for a join RTE (else NULL/zero):`](https://github.com/postgres/postgres/blob/45b9805706fdc726906fc9187c9a7b38c559755c/src/include/nodes/parsenodes.h#L1016) in PG source.

Reply:
I will write my understanding here but this should be taken with a grain of salt.

joinleftcols/joinrightcols seem to refer to columns of the RTE being involved. For example, say that we have `ab(a int, b int)` and when we write the following:

```sql
SELECT * FROM ab first, ab second where first.a = second.a;
```

joinleftcols will be = [1,2] and join rightcols will also be [1,2].

If we drop the first column and do the same:

```sql
ALTER TABLE ab drop column a;
SELECT * FROM ab first, ab second where first.a = second.a;
```

Then joinleftcols and joinrightcols will be [2]. So in case of rte being RELATION_RTE, these fields seem to refer to its original column index ( for example we dropped column `a`, but the original index of column `b` was `2` so joinleftcols is [2] not [1].) 

So in case of RTE being RELATION_RTE, it seems that it is expected to fill these columns with the original indexes of columns, which is the intention of this PR/fix.

When we do the following:

```sql
ALTER TABLE ab drop column a;
SELECT * FROM ab first join ab second on first.b=second.b join ab third on third.b = first.b;
```

For the first join since both sides will be relations these fields will be:

```
      :joinleftcols (i 2)
      :joinrightcols (i 2)
```

After doing the first join, the left side is no longer a relation(JOIN_RTE) but the right side is (ab third) so now the fields are:

```
     :joinleftcols (i 1 2) -- 1,2 refers to the columns of JOIN_RTE(b,b)
     :joinrightcols (i 2) -- 2 refers to the column index of relation ab
```

Reply:
Also it seems that this is how postgres is filling these fields:

https://github.com/postgres/postgres/blob/7fc1a81e4982b6722405b032e21be38dbbef5b80/src/backend/parser/parse_clause.c#L281-L286